File: submodules/mo_grippers/src/grippers/dh-gripper.cpp
#include <algorithm>
#include <plaif/mo-grippers/grippers/dh-gripper.h>

namespace plaif
{
  void DhGripper::initialize(bool should_wait)
  {
    protocol_->send(0x0100, 165);
    if (should_wait)
    {
      std::this_thread::sleep_for(std::chrono::seconds(5));
    }
    std::cout << "DhGripper is initialized." << std::endl;
  }

  void DhGripper::setPositionRaw(int64_t position)
  {
    if (position < LOWER_LIMIT || position > UPPER_LIMIT)
    {
      throw std::out_of_range("Raw position must be between 0 and 1000!");
    }
    protocol_->send(0x0103, position);
  }

  void DhGripper::setPosition(uint16_t scaled)
  {
    if (scaled > 100)
    {
      throw std::out_of_range("Scaled position must be between 0 and 100!");
    }
    int64_t position = static_cast<int64_t>((UPPER_LIMIT - LOWER_LIMIT) / 100.0 * scaled);
    setPositionRaw(position);
  }

  int64_t DhGripper::getPositionRaw()
  {
    return static_cast<uint64_t>(protocol_->receive(0x0202));
  }

  uint16_t DhGripper::getPosition()
  {
    uint64_t raw_value = getPositionRaw();
    raw_value = std::min(static_cast<uint64_t>(UPPER_LIMIT),
                         std::max(static_cast<uint64_t>(LOWER_LIMIT), raw_value));
    return static_cast<uint16_t>(raw_value * 100 / (UPPER_LIMIT - LOWER_LIMIT));
  }

  void DhGripper::setForce(int16_t force)
  {
    if (force < 20 || force > 100)
    {
      throw std::out_of_range("Force must be between 20 and 100!");
    }
    protocol_->send(0x0101, force);
  }

  int16_t DhGripper::getForce()
  {
    return 0;
  }

  int16_t DhGripper::getState()
  {
    return static_cast<int16_t>(protocol_->receive(0x0201));
  }

  void DhGripper::openFingers()
  {
    setPositionRaw(UPPER_LIMIT);
  }

  void DhGripper::closeFingers()
  {
    setPositionRaw(LOWER_LIMIT);
  }
}

File: submodules/mo_grippers/mo_grippers/grippers/dh_gripper.py
import time
from mo_grippers.protocols.modbus import ModbusProtocol
from mo_grippers.grippers.base_gripper import BaseGripper


LOWER_LIMIT = 0
UPPER_LIMIT = 1000


class DhGripper(BaseGripper):
    protocol: ModbusProtocol

    def __init__(self, port="/dev/ttyUSB0", baudrate=115200, slave_id=1):
        super().__init__(
            ModbusProtocol(port=port, baudrate=baudrate, slave_id=slave_id)
        )

    def initialize(self, should_wait=False):
        response = self.protocol.send(0x0100, 165)
        if should_wait:
            time.sleep(5.0)
        print("DhGripper is initialized.")
        return response

    def set_force(self, force):
        if 20 <= force <= 100:
            return self.protocol.send(0x0101, force)
        else:
            raise ValueError("Force must be between 20% and 100%")

    def set_position_raw(self, position):
        assert (
            LOWER_LIMIT <= position <= UPPER_LIMIT
        ), "Position must be between 0 and 1000"
        return self.protocol.send(0x0103, int(position))

    def set_position(self, scaled):
        assert 0 <= scaled <= 100, "Scaled position should be between 0 and 100"
        position = (UPPER_LIMIT - LOWER_LIMIT) / 100 * scaled
        return self.set_position_raw(position)

    def get_position_raw(self):
        return self.protocol.receive(0x0202)

    def get_position(self):
        position = self.get_position_raw()
        return position * 100 / (UPPER_LIMIT - LOWER_LIMIT)

    def get_state(self):
        return self.protocol.receive(0x0201)

    def open_fingers(self):
        return self.set_position_raw(UPPER_LIMIT)

    def close_fingers(self):
        return self.set_position_raw(LOWER_LIMIT)
